staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -1466,6 +1466,29 @@ type ConfigMapProjection struct {
 	Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
 }
 
+// ServiceAccountTokenProjection represents a projected service account token
+// volume. This projection can be used to insert a service account token into
+// the pods runtime filesystem for use against APIs (Kubernetes API Server or
+// otherwise).
+type ServiceAccountTokenProjection struct {
+	// Audience is the intended audience of the token. A recipient of a token
+	// must identify itself with an identifier specified in the audience of the
+	// token, and otherwise should reject the token. The audience defaults to the
+	// identifier of the apiserver.
+	//+optional
+	Audience string `json:"audience,omitempty" protobuf:"bytes,1,rep,name=audience"`
+	// ExpirationSeconds is the requested duration of validity of the service
+	// account token. As the token approaches expiration, the kubelet volume
+	// plugin will proactively rotate the service account token. The kubelet will
+	// start trying to rotate the token if the token is older than 80 percent of
+	// its time to live or if the token is older than 24 hours.Defaults to 1 hour
+	// and must be at least 10 minutes.
+	//+optional
+	ExpirationSeconds int64 `json:"expirationSeconds,omitempty" protobuf:"varint,2,opt,name=expirationSeconds"`
```

Comment:
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required says this should be a pointer

Technically for `Audience` too, but I think we rarely make optional strings pointers... 
```
$ grep -Pzo "(?s)\+optional\n\N*\ \*string\N*\n" types.go | wc -l
14
$ grep -Pzo "(?s)\+optional\n\N*\ string\N*\n" types.go | wc -l
270

$ grep -Pzo "(?s)\+optional\n\N*\ \*int\N*\n" types.go | wc -l
42
$ grep -Pzo "(?s)\+optional\n\N*\ \int\N*\n" types.go | wc -l
38
```

Reply:
So I asked @liggitt about this on slack (so don't fault his chat style response 😄 ) and I wanted to capture his thoughtful perspective:

>I think there are a few things wrapped up in that recommendation... and the fact that it is so inconsistently followed (both in old and new APIs) makes it very hard to find good patterns
>1. we don't want optional omitempty fields getting serialized. for structs, that requires them to be pointers. "empty" string, bool, and int values get omitted without needing to be pointers
>2. when the "empty" value is a valid value and shouldn't be defaulted to something else, we obviously need to make it a pointer
>3. the case you call out (the empty value is not a valid value and we do want to default it to something else) gets weird
>
>if you don't make it a pointer, and default zero-values, it means we'd accept `{"expirationSeconds":0}` and that would mean `{"expirationSeconds":<default value>}`
when from the outside, you'd expect an error about `"expirationSeconds" must be >= 600`.
> as annoying as pointer scaler values are to deal with, for the sake of sane external behavior, we should probably try to differentiate between "explicitly set foo=0" and "no value set"

So he agrees with you. I'll make this a pointer.

Reply:
@awly, we talked about this same thing a few days ago so this opinion might be interesting to you as well ^^